### ISSUE TITLE
Missing types for Microsoft.Web/sites/host and Microsoft.Web/sites/host/functionKeys

### ISSUE THREAD

**Issue by szszoke** [Azure/bicep-types-az#1356]

**Comment by Camios** [comment]
I'm getting **intermittent** Bad Request from attempts with an error message "Encountered an error (InternalServerError) from host runtime."

The two requests which intermittently succeed/fail (sometimes one will work or both will. I don't see a pattern why) are: 

Provisioning operation: `Action`
Type: `Microsoft.Web/sites/host`
Resource: `<function-app-name>/default`

Provisioning operation: `Create`
Type: `Microsoft.Web/sites/host/functionkeys`
Resource: `<function-app-name>/default/<function-host-key-name>`

**Comment by Camios** [comment]
It also seems like ARM template suffers from the same problem

```
{
    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
    "contentVersion": "1.0.0.0",
    "parameters": {},
    "functions": [],
    "variables": {
        "keyName": "apimanagement",
        "FunctionAppName": "func-app"
    },
    "resources": [
        {
            "type": "Microsoft.Web/sites/host/functionKeys",
            "apiVersion": "2022-03-01",
            "name": "[concat(variables('FunctionAppName'), '/default/', variables('keyName'))]",
            "properties": {
                "name": "[variables('keyName')]"
            }
        }

    ],
    "outputs": {}
}
```
The VSCode ARM template says there's a problem with the `"type": "Microsoft.Web/sites/host/functionKeys"`
Specifically `Value must be one of the following values: "Microsoft.Genomics/accounts", "Microsoft.RecoveryServices/vaults", etc`

Works/fails intermittently. When it does fail, here's the deployment error:
```
{
    "status": "Failed",
    "error": {
        "code": "BadRequest",
        "message": "Encountered an error (InternalServerError) from host runtime.",
        "details": [
            {
                "message": "Encountered an error (InternalServerError) from host runtime."
            },
            {
                "code": "BadRequest"
            },
            {}
        ]
    }
}
```

If this affects both bicep and ARM templates, where should I report this problem? Noting one difference is the bicep request seems to make two deployment requests vs one for ARM. It seems fundamentally unsound and that might be why it isn't listed as supported... 

I deployed this ARM template in a phased approach (phase 1 is the initial creation of the function app and some other resources via a bicep template; the second phase publishing the function app). I notice it can intermittently fail when run immediately after phase one or two. But then I can rerun the same script a few minutes later and it worked?
Perhaps I could put this into a retry loop until it works? Or it might just intermittently fail regardless of how close it is run after another phase.

**Comment by dkapadia-jb** [comment]
Has there been any update on this?

**Comment by fleed** [comment]
I have the same problem.
After deploying a function app with bicep, the portal doesn't show any host keys.
Every attempt to generate keys using bicep fails

**Comment by elvismangarae** [comment]
> I'm getting **intermittent** Bad Request from attempts with an error message "Encountered an error (InternalServerError) from host runtime."
> 
> The two requests which intermittently succeed/fail (sometimes one will work or both will succeed or both fail). I don't see a pattern why) are:
> 
> Provisioning operation: `Action` Type: `Microsoft.Web/sites/host` Resource: `<function-app-name>/default`
> 
> Provisioning operation: `Create` Type: `Microsoft.Web/sites/host/functionkeys` Resource: `<function-app-name>/default/<function-host-key-name>`
> 
> What's strange is the the function host key is being created, yet the deploy operations stops. It doesn't seem to get to the next step where I declare an APIM named value resource which attempts a list keys on the new function's host key. In the named value I have a dependsOn the function host key resource.


I ran into this same issue. Issue was resolved by first creating the function app resource, then creating the APIM resource in a module that takes the function app name as a param. That module creates the host key AFTER the intial APIM service is created, by using the apimService.name within the functionapp host key name. I think the way I'm doing it is not important, but just that you need to allow several minutes between app creation and host key creation. For me, creating the APIM service was enough time.

**Comment by CrazyTim** [comment]
@szszoke agreed. Documentation needs to be added for this and the language service updated. This feature does seem half-baked to me, which is probably why the documentation doesn't exist 🤷🏻‍♂️

@Camios I believe the bicep templates actually call the REST APIs under-the-hood, hence why it returns internal server error (500) when the function app is not ready. Our soulution was to poll the funciton app after deployment and wait until it responds. We made a bash script with multiple calls to `az deployment` with the polling in between. Unfortunately we weren't able to get it working in a single bicep template. 

Here's the poll script if anyone is interested:

```bash
#!/usr/bin/env bash
# poll-function-app-host-ready.sh
_resource_group=$1
_function_app_name=$2

echo -e "\nPoll until function app host is ready: '${_resource_group}/${_function_app_name}'"

for i in {1..20}; do
  key=$(az functionapp keys list -g "${_resource_group}" -n "${_function_app_name}" \
         --query "functionKeys.default" -o tsv 2>/dev/null) || true
  if [ -n "$key" ]; then
    echo "Host is ready!"
    #export key
    break
  fi
  echo "Host not ready yet ($i/20); sleeping..."
  sleep 4
done
```
